core/src/plugins/kubernetes/local-mode.ts

```diff
@@ -0,0 +1,711 @@
+          await ensureDir(serviceDirPath)
+          const keyPair = new KeyPair(serviceDirPath)
+          await ProxySshKeystore.generateSshKeys(keyPair, service, log)
+          this.serviceKeyPairs.set(serviceDirPath, keyPair)
+        }
+      })
+    }
+    return this.serviceKeyPairs.get(serviceDirPath)!
+  }
+
+  public shutdown(log: LogEntry): void {
+    this.serviceKeyPairs.forEach((value) => {
+      ProxySshKeystore.deleteFileFailSafe(value.privateKeyPath, log)
+      ProxySshKeystore.deleteFileFailSafe(value.publicKeyPath, log)
+    })
+    this.serviceKeyPairs.clear()
+  }
+}
+
+/*
+ * This can be changed to a "global" registry for all processes,
+ * but now retriable processes are used in local mode only.
+ */
+export class LocalModeProcessRegistry {
+  private readonly retriableProcesses: Map<string, RetriableProcess>
+
+  private constructor() {
+    if (!!LocalModeProcessRegistry.instance) {
+      throw new RuntimeError("Cannot init singleton twice, use LocalModeProcessRegistry.getInstance()", {})
+    }
+    this.retriableProcesses = new Map<string, RetriableProcess>()
+  }
+
+  private static instance?: LocalModeProcessRegistry = undefined
+
+  public static getInstance(): LocalModeProcessRegistry {
+    if (!LocalModeProcessRegistry.instance) {
+      const newInstance = new LocalModeProcessRegistry()
+      registerCleanupFunction("shutdown-local-mode-process-registry", () => newInstance.shutdown())
+      LocalModeProcessRegistry.instance = newInstance
+    }
+    return LocalModeProcessRegistry.instance
+  }
+
+  public register(process: RetriableProcess): void {
+    const root = process.getTreeRoot()
+    this.retriableProcesses.set(root.command, root)
+  }
+
+  public shutdown(): void {
+    this.retriableProcesses.forEach((process) => process.stopAll())
+    this.retriableProcesses.clear()
+  }
+}
+
+export class LocalModeSshPortRegistry {
+  private readonly localsshPorts: Set<number>
+
+  private constructor() {
+    if (!!LocalModeSshPortRegistry.instance) {
+      throw new RuntimeError("Cannot init singleton twice, use LocalModeSshPortRegistry.getInstance()", {})
+    }
+    this.localsshPorts = new Set<number>()
+  }
+
+  private static instance?: LocalModeSshPortRegistry = undefined
+
+  public static getInstance(log: LogEntry): LocalModeSshPortRegistry {
+    if (!LocalModeSshPortRegistry.instance) {
+      const newInstance = new LocalModeSshPortRegistry()
+      registerCleanupFunction("shutdown-local-mode-ssh-port-registry", () => newInstance.shutdown(log))
+      LocalModeSshPortRegistry.instance = newInstance
+    }
+    return LocalModeSshPortRegistry.instance
+  }
+
+  private static cleanupKnownHosts(localPort: number, log: LogEntry): void {
+    const localhostEscaped = localhost.split(".").join("\\.")
+    const command = `sed -i -r '/^\\[${localhostEscaped}\\]:${localPort}/d' \${HOME}/.ssh/known_hosts`
+    try {
+      log.debug("Cleaning up .ssh/known_hosts file...")
```

Comment:
This message would kinda freak me out, thinking the known_hosts file might have been wiped out. Perhaps something like

```suggestion
      log.debug("Cleaning temporary entries from .ssh/known_hosts file...")
```